### PR TITLE
feat(rust): dependency tree and line numbers support for cargo lock file (#179)

### PR DESCRIPTION
**Pull request by AndreyLevchenko**

### CODE REVIEW

File: pkg/rust/cargo/parse_test.go [hunk 1]
```diff
@@ -37,27 +72,40 @@ func TestParse(t *testing.T) {
...
 }
+
+func sortLibs(libs []types.Library) {
+	sort.Slice(libs, func(i, j int) bool {
+		ret := strings.Compare(libs[i].Name, libs[j].Name)
+		if ret == 0 {
+			return libs[i].Version < libs[j].Version
+		}
+		return ret < 0
+	})
```
Comment (DmitriyLewen) [inline pkg/rust/cargo/parse_test.go:104]:
```suggestion
	sort.Slice(libs, func(i, j int) bool {
		return strings.Compare(libs[i].ID, libs[j].ID) < 0
	})
```
looks like we can just use `ID`.

File: pkg/rust/cargo/parse.go [hunk 2]
```diff
@@ -9,13 +12,14 @@ import (
...
+}
 type Lockfile struct {
-	Packages []struct {
-		Name         string   `toml:"name"`
-		Version      string   `toml:"version"`
-		Source       string   `toml:"source,omitempty"`
-		Dependencies []string `toml:"dependencies,omitempty"`
-	} `toml:"package"`
+	Packages []cargoPkg `toml:"package"`
 	Metadata interface{}
```
Comment (DmitriyLewen) [inline pkg/rust/cargo/parse.go:23]: If we still use `toml` decoder:

we don't use tag for `Metadata`, but toml decoder still parses this field:
image
Maybe it make sense to remove `Metadata`?

File: pkg/rust/cargo/parse.go [hunk 3]
```diff
@@ -27,18 +31,57 @@ func NewParser() types.Parser {
 
 func (p *Parser) Parse(r dio.ReadSeekerAt) ([]types.Library, []types.Dependency, error) {
 	var lockfile Lockfile
-	if _, err := toml.DecodeReader(r, &lockfile); err != nil {
+	decoder := toml.NewDecoder(r)
+	if _, err := decoder.Decode(&lockfile); err != nil {
 		return nil, nil, xerrors.Errorf("decode error: %w", err)
 	}
```
Comment (DmitriyLewen) [inline pkg/rust/cargo/parse.go:37]: we will parse lock file again(i mean `naivePkgParser`).
Perhaps it make sense to remove this decoding and take all dependency information from `naivePkgParser`?

File: pkg/rust/cargo/parse.go [hunk 4]
```diff
@@ -27,18 +31,57 @@ func NewParser() types.Parser {
...
+		if dep != nil {
+			deps = append(deps, *dep)
+		}
+	}
+	return libs, deps, nil
+}
+func parseDependencies(pkgId string, pkg cargoPkg) *types.Dependency {
+	dependOn := []string{}
+	for _, pkgDep := range pkg.Dependencies {
+		fields := strings.Fields(pkgDep)
```
Comment (DmitriyLewen) [inline pkg/rust/cargo/parse.go:71]: Can you add comment about this field schema?
if i understand correctly it is: `<pkg_name> <pkg_version> (registry+<link>)` , right?

File: pkg/rust/cargo/naive_pkg_parser.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+			if currentPkg.name != "" {
+				pkgId := utils.PackageID(currentPkg.name, currentPkg.version)
+				currentPkg.setEndPositionIfEmpty(lineNum - 1)
+				idx[pkgId] = currentPkg.position
+			}
+			currentPkg = minPkg{}
+			currentPkg.position.start = lineNum
+
+		} else if matched := pkgNameRegexp.MatchString(line); matched {
+			currentPkg.name = propertyValue(line)
```
Comment (DmitriyLewen) [inline pkg/rust/cargo/naive_pkg_parser.go:56]: There are times when a simple regular expression leaks memory.
What if we remove all `regexp` and use `strings` package?
Something like this:
```suggestion
		} else if strings.HasPrefix(strings.TrimSpace(line), "name") {
			currentPkg.name = propertyValue(line)
```

**Review by DmitriyLewen — APPROVED** [review]
Good work! 
Thank you.

File: pkg/rust/cargo/testdata/cargo_invalid.lock [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+dependencies = [
+ "libc 0.2.54 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
+name = "typemap"
+version = "0.3.3"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "unsafe-any 0.4.2 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (knqyf263) [inline pkg/rust/cargo/testdata/cargo_invalid.lock:20]: `unsafe-any` is not on the package list. Is it possible?

Comment (knqyf263) [reply pkg/rust/cargo/testdata/cargo_invalid.lock:20]: OK, this is an invalid case, but how about `pkg/rust/cargo/testdata/cargo_normal.lock` and `pkg/rust/cargo/testdata/cargo_mixed.lock`? `unsafe-any` is in dependencies, but I didn't find it in package list. I want to use a valid cargo.lock for valid cases.

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**